### PR TITLE
Replicated Sharding improvements

### PR DESCRIPTION
**Pull request by chbatey**
Fixed for when using roles and data centers
Improved test coverage
Introduced ReplicationId separate from PersistenceId

### CODE REVIEW

File: akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/ReplicatedShardingSpec.scala [hunk 1]
```diff
@@ -65,81 +87,197 @@ class ReplicatedShardingSpec
...
+    extends ReplicatedShardingSpec(Role, ReplicatedShardingSpec.roleAConfig, ReplicatedShardingSpec.roleBConfig)
+class DataCenterReplicatedShardingSpec
+    extends ReplicatedShardingSpec(DataCenter, ReplicatedShardingSpec.dcAConfig, ReplicatedShardingSpec.dcBConfig)
+
+abstract class ReplicatedShardingSpec(replicationType: ReplicationType, configA: Config, configB: Config)
+    extends ScalaTestWithActorTestKit(configA)
+    with AnyWordSpecLike
+    with LogCapturing {
+
+  val system2 = ActorSystem(Behaviors.ignore[Any], name = system.name, config = configB)
```
Comment (patriknw) [inline akka-cluster-sharding-typed/src/test/scala/akka/cluster/sharding/typed/ReplicatedShardingSpec.scala:229]: leak, this system is not shutdown

File: akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/ReplicatedEntityProvider.scala [hunk 2]
```diff
@@ -28,29 +28,31 @@ object ReplicatedEntityProvider {
...
    */
-  def apply[M: ClassTag, E](allReplicaIds: Set[ReplicaId])(
+  def apply[M: ClassTag, E](typeName: String, allReplicaIds: Set[ReplicaId])(
       settingsPerReplicaFactory: (EntityTypeKey[M], ReplicaId, Set[ReplicaId]) => ReplicatedEntity[M, E])
       : ReplicatedEntityProvider[M, E] = {
     new ReplicatedEntityProvider(allReplicaIds.map { replicaId =>
-      val typeKey = EntityTypeKey[M](replicaId.id)
-      settingsPerReplicaFactory(typeKey, replicaId, allReplicaIds)
+      // TODO validate it does't contain the separator
+      val typeKey = EntityTypeKey[M](s"$typeName|${replicaId.id}")
```
Comment (patriknw) [inline akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/ReplicatedEntityProvider.scala:54]: use `ReplicationId.Separator`

File: akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/ReplicatedEntityProvider.scala [hunk 3]
```diff
@@ -28,29 +28,31 @@ object ReplicatedEntityProvider {
...
   /**
    * Scala API:
-   *
+   * @param typeName The type name used in the [[EntityTypeKey]]
    * @tparam M The type of messages the replicated entity accepts
    * @tparam E The type for envelopes used for sending `M`s over sharding
    */
-  def apply[M: ClassTag, E](allReplicaIds: Set[ReplicaId])(
+  def apply[M: ClassTag, E](typeName: String, allReplicaIds: Set[ReplicaId])(
       settingsPerReplicaFactory: (EntityTypeKey[M], ReplicaId, Set[ReplicaId]) => ReplicatedEntity[M, E])
```
Comment (patriknw) [inline akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/ReplicatedEntityProvider.scala:50]: why is `allReplicaIds` needed as the 3rd parameter of the `settingsPerReplicaFactory` function? Isn't that already known by the user?

**Review by patriknw — APPROVED** [review]
LGTM, after some minors

File: akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/ReplicatedEntityProvider.scala (hunk 3 above)
Comment (chbatey) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/ReplicatedEntityProvider.scala:50]: 👍 typically a constant so let's remove it

**Pull request merged**